### PR TITLE
Return the default value for a bandit flag if no actions. Cast provided categorical attributes to strings.

### PR DESCRIPTION
**Pull request by lisaah**
[//]: #  (Link to the issue corresponding to this chunk of work)
Fixes: #FF-2573

## Motivation and Context
[//]: #  (Why is this change required? What problem does it solve?)

Tests were modified, catching some edge cases.

- We want to return the default value for a bandit flag with no actions
- Handle dynamic casting for categorical attributes

## Description
[//]: # (Describe your changes in detail)

- Adds check for if there are any actions provided and returns default value if not.
- Converts provided categorical attributes to strings.

## How has this been tested?
[//]: # (Please describe in detail how you tested your changes)
Unit tests.

[//]: # (OPTIONAL)
[//]: # (Add one or multiple labels: enhancement, refactoring, bugfix)

### CODE REVIEW

File: eppo_client/client.py [hunk 1]
```diff
@@ -334,6 +336,10 @@ def get_bandit_action_detail(
             )
             return BanditResult(variation, None)
 
+        # if no actions are given - a valid use case - return the default value
```
Comment (typotter) [inline eppo_client/client.py:339]: We don't want to compute the flag string assignment (because it will log the flag assignment) and then return the default value if no actions are passed. If no actions are passed _and_ the flag is not associated with bandits, we can compute and return the string assignment.
This requires reading the [ `bandits` field of the UFC response](https://github.com/Eppo-exp/sdk-test-data/blob/282589ae9cd5f1382c0f9470618399bade934bc7/ufc/bandit-flags-v1.json#L155).

**Review by typotter — CHANGES_REQUESTED** [review]
Thanks for continuing to tackle the missing features in this SDK, Lisa.

File: eppo_client/client.py [hunk 2]
```diff
@@ -261,13 +261,15 @@ def get_bandit_action(
             actions (ActionContexts | ActionAttributes): The dictionary that maps action keys
                 to their context of actions with their contexts.
                 If supplying an ActionAttributes, it gets converted to an ActionContexts instance.
-            default (str): The default variation to use if the subject is not part of the bandit.
+            default (str): The default variation to use if an error is encountered retrieving the
```
Comment (aarsilv) [inline eppo_client/client.py:264]: Corrected the comment

File: eppo_client/client.py [hunk 3]
```diff
@@ -261,13 +261,15 @@ def get_bandit_action(
...
+                assigned variation.
 
         Returns:
             BanditResult: The result containing either the bandit action if the subject is part of the bandit,
                           or the assignment if they are not. The BanditResult includes:
                           - variation (str): The assignment key indicating the subject's variation.
-                          - action (str): The key of the selected action if the subject is part of the bandit.
+                          - action (str | null): The key of the selected action if the subject was assigned one
```
Comment (aarsilv) [inline eppo_client/client.py:271]: returned actions can be None

File: eppo_client/client.py [hunk 4]
```diff
@@ -286,102 +288,112 @@ def get_bandit_action(
...
+            do_action(result.variation)
         else:
-            do_action(result.action)
+            do_status_quo()
         """
+        variation = default
+        action = None
         try:
-            return self.get_bandit_action_detail(
+            subject_attributes = convert_context_attributes_to_attributes(subject_context)
```
Comment (aarsilv) [inline eppo_client/client.py:299]: `get_bandit_action_detail` was ensuring the attributes were passed in with context and then disaggregating them back down to `Attributes.`

Now that we pass these in before we try to ensure context, we want to do the opposite, which is to make sure they are disaggregated.

File: eppo_client/client.py [hunk 5]
```diff
@@ -286,102 +288,112 @@ def get_bandit_action(
...
+                )
         except Exception as e:
             if self.__is_graceful_mode:
                 logger.error("[Eppo SDK] Error getting bandit action: " + str(e))
-                return BanditResult(default, None)
-            raise e
+            else:
+                raise e
+
+        return BanditResult(variation, action)
```
Comment (aarsilv) [inline eppo_client/client.py:324]: By having this be the only return statement, if we got a variation but hit an error later on evaluating the bandit, we'll still return the variation (to match our logging of it)

File: eppo_client/client.py [hunk 6]
```diff
@@ -286,102 +288,112 @@ def get_bandit_action(
...
 
-    def get_bandit_action_detail(
+    def evaluate_bandit_action(
```
Comment (aarsilv) [inline eppo_client/client.py:326]: Moved assignment getting upstream, so now this method's only responsibility is to get and log the selected action

File: eppo_client/client.py [hunk 7]
```diff
@@ -286,102 +288,112 @@ def get_bandit_action(
...
-            ),
-            "actionCategoricalAttributes": (
-                evaluation.action_attributes.categorical_attributes
-                if evaluation.action_attributes
-                else {}
-            ),
-            "metaData": {"sdkLanguage": "python", "sdkVersion": __version__},
-        }
-        self.__assignment_logger.log_bandit_action(bandit_event)
+        try:
```
Comment (aarsilv) [inline eppo_client/client.py:360]: We didn't have a try-catch surrounding bandit logging, so putting this in place. This way, if the (customer-provided) bandit log method fails, the assignment is still recorded.

File: eppo_client/client.py [hunk 8]
```diff
@@ -405,6 +417,9 @@ def get_bandit_keys(self):
         This can be useful to debug the initialization process.
         """
         return self.__config_requestor.get_bandit_keys()
+    
+    def set_is_graceful_mode(self, is_graceful_mode: bool):
```
Comment (aarsilv) [inline eppo_client/client.py:421]: useful to be able to toggle this outside of client instantiation

File: eppo_client/client.py [hunk 9]
```diff
@@ -442,19 +457,34 @@ def check_value_type_match(
...
+    subject_context: Union[ContextAttributes, Attributes]
+) -> Attributes:
+    if isinstance(subject_context, dict):
+        return subject_context
+    
+    return subject_context.numeric_attributes | subject_context.categorical_attributes
+
 
-def convert_subject_context_to_attributes(
+def convert_subject_context_to_context_attributes(
```
Comment (aarsilv) [inline eppo_client/client.py]: more appropriate name

File: eppo_client/client.py [hunk 10]
```diff
@@ -442,19 +457,34 @@ def check_value_type_match(
...
     subject_context: Union[ContextAttributes, Attributes]
 ) -> ContextAttributes:
     if isinstance(subject_context, dict):
         return ContextAttributes.from_dict(subject_context)
-    return subject_context
+
+    stringified_categorical_attributes = {
+        key: str(value) for key, value in subject_context.categorical_attributes.items()
+    }
```
Comment (aarsilv) [inline eppo_client/client.py:477]: @lisaah's original fix, which is to stringify the values of all explicitly passed in categorical attributes so they can be evaluated correctly against the model

File: test/client_bandit_test.py [hunk 11]
```diff
@@ -98,12 +106,45 @@ def test_get_bandit_action_bandit_does_not_exist():
     assert result == BanditResult("default_variation", None)
 
 
-def test_get_bandit_action_flag_without_bandit():
+def test_get_bandit_action_flag_has_no_bandit():
```
Comment (aarsilv) [inline test/client_bandit_test.py:109]: have this use a real flag and get assigned a non-default (but non-bandit) variation

File: test/client_bandit_test.py [hunk 12]
```diff
@@ -80,16 +81,23 @@ def init_fixture():
...
+    mock_assignment_logger.assignment_events.clear()
+    mock_assignment_logger.bandit_events.clear()
 
 def test_is_initialized():
     client = get_instance()
     assert client.is_initialized(), "Client should be initialized"
 
 
-def test_get_bandit_action_bandit_does_not_exist():
+def test_get_bandit_action_flag_not_exist():
```
Comment (aarsilv) [inline test/client_bandit_test.py:97]: this is really what the test is testing

File: test/client_bandit_test.py [hunk 13]
```diff
@@ -98,12 +106,45 @@ def test_get_bandit_action_bandit_does_not_exist():
...
     client = get_instance()
     result = client.get_bandit_action(
-        "a_flag", "subject_key", DEFAULT_SUBJECT_ATTRIBUTES, {}, "default_variation"
+        "non_bandit_flag", "subject_key", DEFAULT_SUBJECT_ATTRIBUTES, {}, "default_variation"
     )
-    assert result == BanditResult("default_variation", None)
+    assert result == BanditResult("control", None)
+
+@patch.object(BanditEvaluator, 'evaluate_bandit', side_effect=Exception("Mocked Exception"))
+def test_get_bandit_action_bandit_error(mock_bandit_evaluator):
```
Comment (aarsilv) [inline test/client_bandit_test.py:117]: Tests save lives! (part I)

File: test/client_bandit_test.py [hunk 14]
```diff
@@ -162,6 +203,32 @@ def test_get_bandit_action_with_subject_attributes():
         == chosen_action.categorical_attributes
     )
 
+@patch.object(MockAssignmentLogger, 'log_bandit_action', side_effect=Exception("Mocked Exception"))
+def test_get_bandit_action_bandit_logger_error(patched_mock_assignment_logger):
```
Comment (aarsilv) [inline test/client_bandit_test.py:207]: Tests save lives! (part II)

File: eppo_client/client.py (hunk 1 above)
Comment (aarsilv) [reply eppo_client/client.py]: This comment was correct at the time, but has since changed. See this [Slack thread](https://eppo-group.slack.com/archives/C05R6BK5BB9/p1721149341928549).

I've implemented the new plan, and have re-requesting a review

File: eppo_client/client.py (hunk 3 above)
Comment (schmit) [reply eppo_client/client.py:271]: Nit: should say `None` instead of `null`, or better: `Optional[str]` to keep it Pythonic

File: eppo_client/client.py (hunk 10 above)
Comment (schmit) [reply eppo_client/client.py:477]: Should we also ensure conversion to string when subject_context is a dictionary?

Comment (schmit) [reply eppo_client/client.py:477]: I guess in that case we don't know what values should be converted, so bugs can still arise but we have no way to fix.

**Review by schmit — APPROVED** [review]
Thanks for the changes, this looks good to me. I don't feel great that we can have 'null' actions after working so hard to removing the null variation in the UFC, but I also don't know how to better handle the empty actions list case.

File: eppo_client/client.py (hunk 10 above)
Comment (aarsilv) [reply eppo_client/client.py:483]: `ContextAttributes.from_dict()` handles this... bucketing all numeric values numeric attributes and then the string values of strings and bools as categorical attributes

**Review by typotter — APPROVED** [review]
